src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingGraphInterface.java

```diff
@@ -0,0 +1,1063 @@
+        // note that argument testing is taken care of in SequenceForKmers
+        Utils.validate(!alreadyBuilt, "Attempting to add sequence to a graph that has already been built");
+
+        // get the list of sequences for this sample
+        List<SequenceForKmers> sampleSequences = pending.computeIfAbsent(sampleName, s -> new LinkedList<>());
+
+        // add the new sequence to the list of sequences for sample
+        sampleSequences.add(new SequenceForKmers(seqName, sequence, start, stop, count, isRef));
+    }
+
+    /**
+     * Thread sequence seqForKmers through the current graph, updating the graph as appropriate
+     * @param seqForKmers a non-null sequence
+     */
+    protected void threadSequence(final SequenceForKmers seqForKmers) {
+        final int startPos = findStart(seqForKmers);
+        if ( startPos == -1 ) {
+            return;
+        }
+
+        final MultiDeBruijnVertex startingVertex = getOrCreateKmerVertex(seqForKmers.sequence, startPos);
+
+        // increase the counts of all edges incoming into the starting vertex supported by going back in sequence
+        if (INCREASE_COUNTS_BACKWARDS) {
+            increaseCountsInMatchedKmers(seqForKmers, startingVertex, startingVertex.getSequence(), kmerSize - 2);
+        }
+
+        if ( debugGraphTransformations ) {
+            startingVertex.addRead(seqForKmers.name);
+        }
+
+        // keep track of information about the reference source
+        if ( seqForKmers.isRef ) {
+            if ( refSource != null ) {
+                throw new IllegalStateException("Found two refSources! prev: " + refSource + ", new: " + startingVertex);
+            }
+            referencePath = new ArrayList<>(seqForKmers.sequence.length - kmerSize);
+            referencePath.add(startingVertex);
+            refSource = new Kmer(seqForKmers.sequence, seqForKmers.start, kmerSize);
+        }
+
+        // loop over all of the bases in sequence, extending the graph by one base at each point, as appropriate
+        MultiDeBruijnVertex vertex = startingVertex;
+        for ( int i = startPos + 1; i <= seqForKmers.stop - kmerSize; i++ ) {
+            vertex = extendChainByOne(vertex, seqForKmers.sequence, i, seqForKmers.count, seqForKmers.isRef);
+            if ( seqForKmers.isRef ) {
+                 referencePath.add(vertex);
+            }
+            if ( debugGraphTransformations ) {
+                vertex.addRead(seqForKmers.name);
+            }
+        }
+        if ( seqForKmers.isRef ) {
+            referencePath = Collections.unmodifiableList(referencePath);
+        }
+    }
+
+    /**
+     * Changes the threading start location policy.
+     *
+     * @param value  {@code true} if threading will start only at existing vertices in the graph, {@code false} if
+     *  it can start at any unique kmer.
+     */
+    public final void setThreadingStartOnlyAtExistingVertex(final boolean value) {
```

Comment:
I don't love these mutators that are only used for tests.  I don't love adding a million more constructor parameters either though.

Reply:
I think this is the lesser of two evils. Plus that would involve changing a bunch of test code from before my time. Whats even more infuriating is that the argument code that builds the assembly engine for HC and M2 use a mix of constructor parameters and mutators with no real cohesive pattern between the two. Its a mess but its not really clear that there is some solution to all the problems.

Reply:
The typical solution would be to add a factory for it that lets you specify it incrementally and uses private constructor.  It's kind of heavy weight though.  Another option is a parameter object, or a varargs set of options.  Also kind of heavy though.

Reply:
> Whats even more infuriating is that the argument code that builds the assembly engine for HC and M2 use a mix of constructor parameters and mutators with no real cohesive pattern between the two.

So true.  I'll take some blame for noticing it and not doing anything.  I'm hoping that in a few months M2 and HC will just have the same defaults and we can put an end to this all.  All the relevant powers that be are in favor of unifying, so we just have to validate bringing M2 changes over to HC.

Reply:
Should I tackle this issue now? I'm inclined to kick that can down the road until we decide to unify the arguments and revisit this. I agree the factory approach is heavy-handed but its not exactly easy to have the constructor take a parameter object when there are two versions floating around with subtly different behavior between HC and M2, and having them both build some unified object gets us back to square 1 anyway. Plus that would complicate the code relating to testing this assemblers somewhat.

Reply:
I think you can punt on it.